### PR TITLE
Fixes #4262 - Exclude bots, spiders and thread authors from thread views

### LINKED ISSUE #4262: Add options to prevent spiders and thread autor from increasing thread view count
I found a solution here: https://community.mybb.com/post-882217.html

But I ask why is there not an option in the settings to prevent this or even a standard configuration like the one above?

### PR DESCRIPTION
**Pull request by Eldenroot**
Resolves #4262 

Thank you buddy @lairdshaw - for your mentoring, I really appreciate it!

### CODE REVIEW

**Comment by lairdshaw** [comment]
Happy to have helped, buddy. I think you ended up with a good outcome.

File: showthread.php [hunk 1]
```diff
@@ -734,15 +734,32 @@
...
+	if($mybb->settings['threadviewcountexcludeguests'] == 1 && $mybb->user['uid'] == 0)
+	{
+		$count_view = false;
+	}
+	if($mybb->settings['threadviewcountexcludethreadauthor'] == 1 && $mybb->user['uid'] == $thread['uid'])
+	{
+		$count_view = false;
+	}
+	
+	if($count_view == true)
```
Comment (dvz) [inline showthread.php]: It might be better to use a single `if` condition for the block below

**Comment by dvz** [comment]
It may be more intuitive to flip the meaning of these settings and use _include_ instead of _exclude_.

File: showthread.php (hunk 1 above)
Comment (lairdshaw) [reply showthread.php]: Yes, you're right, it might be. @Eldenroot and I were going for an "Easy enough for a total noob to grok" control flow.

**Comment by lairdshaw** [comment]
> It may be more intuitive to flip the meaning of these settings and use _include_ instead of _exclude_.

Doh. It was me who advised the inversion. I still stand by that choice. I think inclusion is more ambiguous in this scenario. e.g., The user is a guest, and the setting says "include guests" so s/he is included, but the user is also *not* a spider, and the setting says "include spiders", so on that basis should s/he _not_ be included...?... It's ambiguous. A principle of exclusion is not (IMO) *as* ambiguous: if *any* of the exclusion criteria are applicable, then the user is excluded. If *none* of them apply, then the user is included.

(I do realise that from a strictly logical point of view, this whole argument could be turned around on me, so I guess all I'm saying is that I find exclusion more intuitive - but if the consensus is that it's not, then that's OK with me).

File: showthread.php (hunk 1 above)
Comment (Eldenroot) [reply showthread.php]: @dvz - Done, simplified!

File: showthread.php [hunk 2]
```diff
@@ -734,15 +734,18 @@
 	}
 
 	// Increment the thread view.
-	if($mybb->settings['delayedthreadviews'] == 1)
+	if(!($mybb->settings['threadviewcountexcludespiders'] == 1 && $session->is_spider == true || $mybb->settings['threadviewcountexcludeguests'] == 1 && $mybb->user['uid'] == 0 || $mybb->settings['threadviewcountexcludethreadauthor'] == 1 && $mybb->user['uid'] == $thread['uid']))
```
Comment (dvz) [inline showthread.php]:
```suggestion
	if(!(
		($mybb->settings['threadviewcountexcludespiders'] == 1 && $session->is_spider == true) ||
		($mybb->settings['threadviewcountexcludeguests'] == 1 && $mybb->user['uid'] == 0) ||
		($mybb->settings['threadviewcountexcludethreadauthor'] == 1 && $mybb->user['uid'] == $thread['uid'])
	))
```
Split + parentheses for clarity

**Comment by Eldenroot** [comment]
@dvz thank you, now it should be fine.

**Comment by dvz** [comment]
Currently _No_ values would result in double negatives - skimming through descriptions and spotting `[...] guests [...]` with a _Yes / No_ setting may be interpreted as _include/exclude guests_.

Would `include X` be mistaken for `[only] include X`, rather than `[also] include X`?

We can also use alternatives like _Increment Thread View Count on Guest Visits_.

File: install/resources/settings.xml [hunk 3]
```diff
@@ -876,6 +876,30 @@ min=0]]></optionscode>
...
+			<optionscode><![CDATA[yesno]]></optionscode>
+			<settingvalue><![CDATA[1]]></settingvalue>
+			<isdefault>0</isdefault>
+		</setting>
+		<setting name="threadviewcountexcludeguests">
+			<title>Exclude views by guests in thread view count?</title>
+			<description><![CDATA[Here you can exclude views by guests from the thread view counter.]]></description>
+			<disporder>16</disporder>
+			<optionscode><![CDATA[yesno]]></optionscode>
+			<settingvalue><![CDATA[0]]></settingvalue>
```
Comment (Sama34) [inline install/resources/settings.xml]: Should be On by default.

File: install/resources/settings.xml [hunk 4]
```diff
@@ -876,6 +876,30 @@ min=0]]></optionscode>
...
+			<isdefault>0</isdefault>
+		</setting>
+		<setting name="threadviewcountexcludethreadauthors">
+			<title>Exclude views by thread author in thread view count?</title>
+			<description><![CDATA[Here you can exclude views by thread author from the thread view counter.]]></description>
+			<disporder>17</disporder>
+			<optionscode><![CDATA[yesno]]></optionscode>
+			<settingvalue><![CDATA[0]]></settingvalue>
```
Comment (Sama34) [inline install/resources/settings.xml]: Should be On by default too.

File: install/resources/settings.xml [hunk 5]
```diff
@@ -876,6 +876,30 @@ min=0]]></optionscode>
 			<settingvalue><![CDATA[0]]></settingvalue>
 			<isdefault>1</isdefault>
 		</setting>
+		<setting name="threadviewcountexcludespiders">
+			<title>Exclude views by spiders in thread view count?</title>
+			<description><![CDATA[Here you can exclude views by spiders from the thread view counter.]]></description>
```
Comment (Sama34) [inline install/resources/settings.xml]:
```suggestion
		<setting name="threadviews_countspiders">
			<title>Increment Thread View Count on Spider Visits</title>
			<description><![CDATA[Turn this setting On to increment the thread view count on spiders visits.]]></description>
```

Take this how as guide for the rest. Note the wording as this setting (as per suggestion) should be the one that is Off by default.

**Review by Sama34 — CHANGES_REQUESTED** [review]

**Comment by Sama34** [comment]
I think a _include_ approach with a _Increment Thread View Count on Guest Visits_ wording is the best approach to avoid confusion.

**Comment by Eldenroot** [comment]
@Sama34 - please check this comment from @dvz - https://github.com/mybb/mybb/issues/4262#issuecomment-784451016

It was changed to off by default because of the request.

**Comment by lairdshaw** [comment]
I defer to your stronger position, @dvz and @Sama34. Exclusion was not, as I'd initially thought, as good a framing as inclusion. You guys are correct...

File: install/resources/settings.xml (hunk 5 above)
Comment (Eldenroot) [reply install/resources/settings.xml]: Done!

File: install/resources/settings.xml (hunk 4 above)
Comment (Eldenroot) [reply install/resources/settings.xml]: Done!

File: install/resources/settings.xml (hunk 3 above)
Comment (Eldenroot) [reply install/resources/settings.xml]: Done!

**Comment by Eldenroot** [comment]
Thank you all for your valuable feedback! Now it should be fixed and all your suggestions were implemented.

File: showthread.php [hunk 6]
```diff
@@ -734,15 +734,22 @@
 	}
 
 	// Increment the thread view.
-	if($mybb->settings['delayedthreadviews'] == 1)
+	if(
+		($mybb->settings['threadviews_countspiders'] == 1 && $session->is_spider == true) ||
+		($mybb->settings['threadviews_countguests'] == 1 && $mybb->user['uid'] == 0) ||
+		($mybb->settings['threadviews_countthreadauthor'] == 1 && $mybb->user['uid'] == $thread['uid'])
+	)
```
Comment (dvz) [inline showthread.php]:
```suggestion
	if(
		(
			$mybb->user['uid'] == 0 &&
			(
				($session->is_spider == true && $mybb->settings['threadviews_countspiders'] == 1) ||
				($session->is_spider == false && $mybb->settings['threadviews_countguests'] == 1)
			)
		) ||
		(
			$mybb->user['uid'] != 0 &&
			($mybb->settings['threadviews_countthreadauthor'] == 1 || $mybb->user['uid'] != $thread['uid'])
		)
	)
```

Addressed:
- missed visitors from none of the groups from settings
- confusing guests and spiders (`uid == 0`)
- confusing guest thread authors with other guests

File: install/resources/settings.xml [hunk 7]
```diff
@@ -876,6 +876,30 @@ min=0]]></optionscode>
 			<settingvalue><![CDATA[0]]></settingvalue>
 			<isdefault>1</isdefault>
 		</setting>
+		<setting name="threadviews_countspiders">
+			<title>Increment Thread View Count on Spider Visits?</title>
+			<description><![CDATA[Here you can include views by spiders in the thread view counter.]]></description>
+			<disporder>15</disporder>
+			<optionscode><![CDATA[yesno]]></optionscode>
+			<settingvalue><![CDATA[0]]></settingvalue>
+			<isdefault>0</isdefault>
```
Comment (dvz) [inline install/resources/settings.xml]: I believe `isdefault` should be `1` for all core settings

Comment (Eldenroot) [reply install/resources/settings.xml]: I will fix in 9 hours. I was confused with this, not properly documented.

Thx

Comment (Eldenroot) [reply install/resources/settings.xml]: Changed.

**Pull request merged**